polyphemus/spec/fixtures/etls/redcap/test2.rb

```diff
@@ -0,0 +1,94 @@
+# Define the new classes dynamically
+#
+# :record, :event, :repeat_instance, :value
+
+define_model("Make").class_eval do
+  def identifier(record_name)
+    record_name
+  end
+end
+
+define_model("Model").class_eval do
+  def identifier(record_name, event_name)
+    event_name
+  end
+end
+
+define_model("Year").class_eval do
+  def identifier(record_name, event_name, (repeat_instrument, repeat_id) )
+    [ record_name, event_name, repeat_id ].join(' ')
+  end
+end
+
+define_model("Feature").class_eval do
+  def patch(magma_record_name, record)
+    record[:year] = magma_record_name.split('-').values_at(1,2,4).join(' ')
+  end
+end
+
+define_model("Award").class_eval do
+  def patch(magma_record_name, record)
+    record[:model] = magma_record_name.split('-')[2]
+  end
+end
+
+def config
+  {
+    models: {
+      make: {
+        each: [ "record" ],
+        scripts: [
+          {
+            attributes: {
+              date_of_founding: "dof"
+            }
+          }
+        ]
+      },
+      model: {
+        each: [ "record", "event" ],
+        scripts: [
+          {
+            attributes: {
+              type: "car_class"
+            }
+          }
+        ]
+      },
+      year: {
+        each: [ "record", "event", "repeat" ],
```

Comment:
Can you explain a bit this new notion in the project config? I see three versions here, not clear what the expected outcome would be or the syntax differences?

Reply:
Apologies - there are two main introductions in this syntax:

1) The `each` Array defines the items in the redcap record that will be used to group a set of redcap records into magma records. Each term in the each Array corresponds to an iterable property of the redcap record - this doesn't exactly map cleanly to a single field in the record, so there is some abstraction here (e.g., "repeat" encompasses several fields). When trying to build Magma records, the loader will extract each of the values in the `each` Array from every redcap record and group according to the corresponding key of strings. These values are also passed in to the model's `identifier` method to construct the Magma identifier from the various redcap values.

2. The entities can be filtered according to the identifier values in order to restrict or exclude certain records. If the `each` term is a single-value hash instead of a string, the hash value is used to filter the key (e.g. `each: [ "record", { "entity" => /Entity [123]/ } ]`).